### ISSUE TITLE
`sls deploy --package` deletes .serverless directory

### ISSUE THREAD

**Issue by atz** [serverless/serverless#7298]
For some reason, the `deploy --package` option causes serverless to delete the `.serverless` directory, even when nothing has changed or gets deployed.  This behavior is unexpected, because `sls deploy` does not nuke the directory.

In particular, this is bad when a specially built `.serverless` directory is built, intending to be deployed by serverless, and the command is:
```
sls deploy --package .serverless
```
In that case, the specially built zipfiles are also deleted along with the directory!  

The more common case, comparing `sls deploy` and `sls deploy --package temp` (with a copy of `.serverless` dir:
```
$ ls .serverless/
cloudformation-template-update-stack.json	dumblogger.zip					serverless-state.json				signer.zip					wrapper.zip
$ sls deploy --verbose
Serverless: Packaging service...
Serverless: Excluding development dependencies...
Serverless: Excluding development dependencies...
Serverless: Excluding development dependencies...
Serverless: Service files not changed. Skipping deployment...
Service Information
service: signer
stage: dev
region: us-east-1
stack: signer-dev
api keys:
  None
endpoints:
  None
functions:
  dumblogger: signer-dev-dumblogger
  wrapper: signer-dev-wrapper
  signer: signer-dev-signer
layers:
  None

Stack Outputs...
ServerlessDeploymentBucketName: ...

$ ls .serverless/
cloudformation-template-update-stack.json	dumblogger.zip					serverless-state.json				signer.zip					wrapper.zip
$ cp -R .serverless temp
$ sls deploy --verbose --package temp
Serverless: Service files not changed. Skipping deployment...
Service Information
service: signer
stage: dev
region: us-east-1
stack: signer-dev
api keys:
  None
endpoints:
  None
functions:
  dumblogger: signer-dev-dumblogger
  wrapper: signer-dev-wrapper
  signer: signer-dev-signer
layers:
  None

Stack Outputs...
ServerlessDeploymentBucketName: ...

$ ls .serverless/
ls: .serverless/: No such file or directory
```

Expected behavior: deployment never deletes the `.serverless` directory, regardless of options used.

**Comment by medikoo** [comment]
> This behavior is unexpected, because sls deploy does not nuke the directory.

Actually `sls deploy` (without `--package` path) also nukes the `.serverless` directory, and recreates the content of it.

Still, it's true that with `--package` param, there's no reason for serverless to remove `.serverless` directory.

We're open for PR that improves that.

**Comment by rdsedmundo** [comment]
The `.serverless` directory is a "temporary working directory" of the Serverless framework, and it's always used, even when you pass the `--path (-p)` property to the package command. What you're thinking is that the old content from your previous package is getting deleted at the end, but in reality, it is at the beginning, always. The following is what happens (on <ins>every</ins> packaging):
1) .serverless is deleted.
2) .serverless is recreated, the files are placed in there.
3) **IF** the `-p` option is provided, the contents of `.serverless` is copied over to the `-p` argument directory, then the `.serverless` is deleted. 

I don't think the work to change this behavior is worth the benefit, considering that you can simply use the `-p` option always and have the content preserved (as long as you don't repeat it further to another packaging). The `.serverless` should be considered as a temporary directory and not something to rely on.

**Comment by medikoo** [comment]
@rdsedmundo note request questions behavior of `sls deploy` (not `sls package`), which is more like:

 1. .serverless is deleted.
 1.  `.serverless` is recreated, the files are placed in there (If the `-p` option is provided, then instead of generating content of .`serverless`, contents of  `-p` argument directory is copied to `.serverless`)
1. Deployment happens
1. If `-p` was provided `.serverless` directory is deleted

Technically Framework could just make deployemt in context of  `-p` directory and leave eventually existing `.serverless` folder untouched.

**Comment by keith1024** [comment]
I also feel like the current behaviour is not intuitive. I am happy with the `sls package` command nuking the `.serverless` folder and repackaging the project into it.

But the fact that the directory remains there after it is complete, leads to a logical assumption that I can now call `sls deploy --package .serverless` to deploy the already packaged project without sitting through the packaging functionality.

Instead you're greeted with an error that `.serverless` doesn't exist, because `deploy` just nuked it. That behaviour is perfectly understandable when the `--package` option is omitted, but bizarre when it is provided.

**Comment by s-kikuchi** [comment]
@medikoo and everyone,

I want to fix this issue if there is no one to do it.
However, I'm not confident about my fix plan. So please anyone check it out. 
If it's something wrong, please write it down which part is wrong.

#### sls deploy --p xxx (xxx is not `.serverless`)
- [ ] Fix not to delete `.serverless` directory when executing the command.
- [ ] Fix not to copy the context of `-p` directory into `.serverless` directory when executing the command.

#### sls deploy --p .serverless
- [ ] Fix so that there are no errors when executing the command.

#### sls package (no options)
- [ ] Enhance to package the project into `.serverless` when executing the command.

Thank you.

**Comment by medikoo** [comment]
@s-kikuchi I believe it should be:

#### sls deploy --p xxx (xxx is whatever, could be `.serverless` or not)
- [ ] Fix not to copy the context of `-p` directory into `.serverless` directory when executing the command (just deploy directly from provided path)
- [ ] Fix to not delete `.serverless` directory after deployment

And otherwise things should remain working as they work now

**Comment by petergaultney** [comment]
Would really appreciate a fix to this. Because of stack size limitations in CloudFormation, we're forced to split our stacks into many parallel stacks, but can't deploy them in parallel from the same directory when they're all writing to the same temp directory.  Splitting them out into their own directories would require so many symlinks that it's terrifying to even think about.

**Comment by duyetmai** [comment]
so do you know when this issue will be fixed?

**Comment by Graham42** [comment]
@petergaultney for the parallel issue, there a related feature request that might be of interest:  https://github.com/serverless/serverless/issues/5002

**Comment by medikoo** [comment]
I've looked into this issue more deeply, and prepared a detailed spec on how it can be fixed.

It's not straightforward patch, but I did my best to make spec rich in details so it can be done by anyone who's not deeply familiar with Framework internals.

PR's are welcome!

**Comment by medikoo** [comment]
(Spec was edited into main issue description)

**Comment by Graham42** [comment]
Thanks for the spec! I'll aim take a look in the next couple days.

I had a fix for https://github.com/serverless/serverless/issues/5002 that might also address this, but needs work for invokeLocal. If I haven't responded here in the next couple days, feel free to adopt my changes. My forked branch is here: https://github.com/Graham42/serverless/pull/1

**Comment by medikoo** [comment]
@Graham42 As I checked, it seems #5002 needs very same solution (whole problem is that Framework always does work in `.serverless` folder).

So I guess best if you compare your idea with spec I proposed and  if you feel spec need adjusting please provide details, any feedback is very welcome

**Comment by Graham42** [comment]
I agree that the framework would ideally use a system temp directory for temporary packaging. However this could be a breaking change for plugins that rely on the compilation, packaging, etc to happen in the `.severless/` directory and hook in at various points to read/write/modify files in that directory.

For backwards compatibility I think we should continue to use the `.serverless/` directory unless some boolean configuration or CLI parameter has be set. (maybe `--use-tmp`?). If enabled, then the framework would provision a temporary directory that is used everywhere in the framework to solve this issue as well as #5002 and #7345. In the next major version of Serverless this flag could be dropped as a breaking change that requires all plugins to make a call to the framework to get the temporary directory instead of assuming `.serverless/`

**Comment by medikoo** [comment]
> This could be a breaking change for plugins that rely on the compilation, packaging, etc to happen in the `.severless/`

@Graham42 that's a good point, still, this looks to me as internal Framework quirk, and what plugins actually should rely on is more the final _package_ directory (which by default is `.serverless`, and it'll stay that way after this refactor).

Do we know plugins which attach to _temporary_ build directory (which now in all cases is `.serverless`) before Framework ensures files are at designated _package_ directory?

**Comment by adminy** [comment]
why is this still an issue forever much later?

**Comment by dashmug** [comment]
> why is this still an issue forever much later?

You can help by submitting a PR or suggesting fixes.

**Comment by adminy** [comment]
Gladly @dashmug if someone can explain why this was created https://github.com/serverless/serverless/blob/45cbd525014ad3bf60aac3d482ec76ce68b51820/lib/plugins/aws/common/lib/cleanup-temp-dir.js#L7-L12


like what makes .serverless a temp directory? Isn't this supposed to be your build artifacts directory?
